fetcher.go

```diff
@@ -0,0 +1,173 @@
+package fetcher
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/johngibb/fetcher/lru"
+)
+
+// batch represents a single fetch.
+type batch struct {
+	ids     map[int64]struct{} // set of ids in batch
+	runOnce sync.Once          // ensure batch is only fetched once
+	done    chan struct{}      // signal that batch is done
+
+	results map[int64]interface{} // results from batch
+	err     error                 // error from batch
+}
+
+func newBatch() *batch {
+	return &batch{
+		done:    make(chan struct{}),
+		ids:     make(map[int64]struct{}),
+		results: make(map[int64]interface{}),
+	}
+}
+
+type cacheEntry struct {
+	val     interface{}
+	expires time.Time
+}
+
+// Fetcher caches and batches requests.
+type Fetcher struct {
+	wait      time.Duration                        // max time to wait for batch to fill up
+	batchSize int                                  // once batch exceeds this size, it's always run
+	fetch     func([]int64) ([]interface{}, error) // fetch function
+	cacheTTL  time.Duration                        // cache TTL; zero disables caching.
+
+	batch   *batch       // current pending batch
+	cache   *lru.Cache   // response cache
+	muBatch sync.Mutex   // synchronizes access to batch
+	muCache sync.RWMutex // synchronizes access to cache
+}
+
+func New(wait, cacheTTL time.Duration, batchSize int, fetch func([]int64) ([]interface{}, error)) *Fetcher {
+	return &Fetcher{
+		batchSize: batchSize,
+		cacheTTL:  cacheTTL,
+		wait:      wait,
+		fetch:     fetch,
+		batch:     newBatch(),
+		cache:     lru.New(10),
+	}
+}
+
+func (f *Fetcher) submit(ids []int64) *batch {
+	// Add the ids to the current batch.
+	f.muBatch.Lock()
+	batch := f.batch
+	for _, id := range ids {
+		batch.ids[id] = struct{}{}
+	}
+	currentBatchSize := len(batch.ids)
+	f.muBatch.Unlock()
+
+	// If the batch exceeds the batch size, run it now.
+	if currentBatchSize >= f.batchSize {
+		f.run(batch)
+		return batch
+	}
+
+	// Otherwise, start a timer to run it after f.wait.
+	go func() {
+		<-time.After(f.wait)
+		f.run(batch)
+	}()
+
+	return batch
+}
+
+func (f *Fetcher) run(batch *batch) {
+	batch.runOnce.Do(func() {
+		// Start a new batch for the next request.
+		f.muBatch.Lock()
+		f.batch = newBatch()
+		f.muBatch.Unlock()
+
+		// Reassemble the id set into a slice of ids.
+		ids := make([]int64, 0, len(batch.ids))
+		for id, _ := range batch.ids {
+			ids = append(ids, id)
+		}
+
+		// Perform the fetch.
+		results, err := f.fetch(ids)
```

Comment:
Is it ok that len(ids) might be significantly greater than the desired batch size, possibly even by several multiples? (Unlikely scenario, maybe, but many goroutines might separately populate a single batch with disjoint sets of IDs.)

Or is it the responsibility of the "fetch" function to handle that case?

Reply:
I thought about that, and reasoned that it's fine as-is (but probably should be documented as a caveat).

Rationale: the `batchSize` should be larger than the typical size of `ids` being requested in a call to `Get`, or else the batching is ineffectual. Given that, a batch in a single fetch can only ever contain `batchSize + len(mostRecentRequest.ids)`, which _should_ mean it's between `[batchSize, 2*batchSize]`. For our use case, of fetching ~500 restaurants on average, or maybe ~1500 in worst case, I'd set a batch size of 1000 and be fine with occasional queries pulling 1500 (since, that's what's happening now anyway).

If it was simple to enforce the batch size, I'd be open to it, but I think it'd add a lot of complexity (since now a request may be waiting on multiple batches, and potentially getting back multiple errors).

What do you think?

Reply:
re:
> possibly even by several multiples
it couldn't exceed `batchSize` by several multiples unless a single call to `Get` had a massive list of ids, right? or am I missing something?

I think the [code that executes the current batch immediately as soon as it gets oversized](https://github.com/johngibb/fetcher/pull/1/files#diff-b21be20702219cd7ad6a3288b7655a84R68) ensures that only one request is allowed to overflow the batch size; there's no possibility that several goroutines could push it over by their aggregate size of requests.

Reply:
You're no longer in the critical section at that point, so I believe another goroutine could call Fetcher.submit, acquire f.muBatch, and add more keys to f.batch.

Reply:
That said, I do think it's unlikely that the query would _greatly_ exceed the target batch size, since the submitted sets of ID are probably going to have a lot of overlap, and it doesn't look like much is happening between releasing the mutex and reacquiring it inside of the sync.Once

Reply:
BTW, I wouldn't implement the cap on batch size by creating multiple batches and waiting on multiple channels. I would just call the "fetch" function multiple times and compiling the results (bailing out and discarded all results should it return an error).